### DISCUSSION TITLE
Restarting GUI after closing window/EventLoop (re-)creation

### DISCUSSION THREAD

**Discussion by bobmoretti** [emilk/egui#2151] [Q&A] [answered]
I have a use case where my eframe gui code lives in a Windows DLL. My use case will never run on any other platform -- it's tied to win32.

A third party application calls a `start()` and `stop()` export from my DLL. In the `start()` I want to spawn a GUI. Ideally, the `stop()` function will close the GUI window, but at the very least I would like subsequent calls of the `start()` function to show the GUI again if the user closed the main window.

With the advent of  `EventLoopBuilder` in the native options in eframe master, I was able to use the `::any_thread()` option that lets me start a winit event loop that can run in an arbitrary thread.

However, when I try to create the GUI a second time, I get the following panic
```
(3e68) ERROR  thread '<unnamed>' panicked at 'Creating EventLoop multiple times is not supported.'
```
I guess a winit limitation is that you only get _one chance_ to create an event loop (internal implementation appears to use a `OnceCell`)? It's not like I want multiple simultaneous event loops; I just want to create another one when the original is dropped.

Is there any straightforward workaround to this situation?

**Comment by emilk** [comment]
One workaround is to only create the event loop once and keep it around for the next window

**Reply by bobmoretti** [reply]
Thanks, agreed, that makes sense. Is there a way to supply my own event loop to eframe? I didn't see one but I might have missed something.

If not, would you be interested in a patch that does this?

**Comment by bobmoretti** [comment, marked as answer]
Digging in a bit further, it looks like this workaround is what eframe is already trying to do. From `eframe/src/native/run.rs`:
```rust
/// Access a thread-local event loop.
///
/// We reuse the event-loop so we can support closing and opening an eframe window
/// multiple times. This is just a limitation of winit.
fn with_event_loop(
    mut native_options: epi::NativeOptions,
    f: impl FnOnce(&mut EventLoop<RequestRepaintEvent>, NativeOptions),
) {
    use std::cell::RefCell;
    thread_local!(static EVENT_LOOP: RefCell<Option<EventLoop<RequestRepaintEvent>>> = RefCell::new(None));

    EVENT_LOOP.with(|event_loop| {
        // Since we want to reference NativeOptions when creating the EventLoop we can't
        // do that as part of the lazy thread local storage initialization and so we instead
        // create the event loop lazily here
        let mut event_loop = event_loop.borrow_mut();
        let event_loop = event_loop
            .get_or_insert_with(|| create_event_loop_builder(&mut native_options).build());
        f(event_loop, native_options);
    });
}
```
This code creates a single event loop _per thread_. My code was terminating the thread and just spawning a new one every time it needed to recreate the GUI. That arrangement was very simple from an implementation standpoint. As a workaround, I can just keep the same thread around and send a message to it every time I need to restart the GUI.
